odoo/models.py

```diff
@@ -6350,25 +6362,34 @@ def process(model, id_vals):
                     for key in self.pool.field_depends_context[field]
                 )
                 model = self.env(context=context_none)[model_name]
-                id_vals = defaultdict(dict)
+                # consume all record ids from all dirty fields of the model
+                fields_recs = dict()
                 for field in model._fields.values():
-                    ids = self.env.cache.clear_dirty_field(field)
-                    if not ids:
-                        continue
-                    records = model.browse(ids)
-                    values = list(self.env.cache.get_values(records, field))
-                    assert len(values) == len(records), \
-                        f"Could not find all values of {field} to flush them\n" \
-                        f"    Context: {self.env.context}\n" \
-                        f"    Cache: {self.env.cache!r}"
-                    for record, value in zip(records, values):
-                        if not field.translate:
-                            value = field.convert_to_write(value, record)
-                            value = field.convert_to_column(value, record)
-                        else:
-                            value = field._convert_from_cache_to_column(value)
-                        id_vals[record.id][field.name] = value
-                process(model, id_vals)
+                    if (ids := self.env.cache.clear_dirty_field(field)):
+                        # set.pop() used in a loop is in O(1) while OrderedSet.pop() is in O(n),
+                        # the ids are subject to `sorted()` later in _write() anyway.
+                        fields_recs[field] = set(ids)
+                if not fields_recs:
+                    continue
+                # re-arrange cache data into record updates batched by same fields and values,
+                # while consuming it to keep space complexity low
+                updates = defaultdict(list)
+                while (field := next(iter(fields_recs), None)):
+                    ids = fields_recs.pop(field)
+                    while ids:
+                        id = ids.pop()
+                        record = model.browse(id)
+                        vals = {field.name: get_cache_value(field, record)}
+                        for other_field in fields_recs:
+                            try:
+                                fields_recs[other_field].remove(id)
+                            except KeyError:
+                                continue
+                            vals[other_field.name] = get_cache_value(other_field, record)
+                        updates[frozendict(vals)].append(id)
+                    # prune drained id sets
+                    fields_recs = {f: fields_recs[f] for f in fields_recs if not fields_recs[f]}
```

Comment:
```suggestion
                while fields_recs:
                    field, ids = fields_recs.popitem()
                    while ids:
                        id_ = ids.pop()
                        record = model.browse(id_)
                        vals = {field.name: get_cache_value(field, record)}
                        for other_field in fields_recs:
                            try:
                                fields_recs[other_field].remove(id_)
                            except KeyError:
                                continue
                            vals[other_field.name] = get_cache_value(other_field, record)
                        updates[frozendict(vals)].append(id_)
                    # prune drained id sets
                    for f in fields_recs.keys():
                        if not fields_recs[f]:
                            del fields_recs[f]
```

Multiple suggestions:

* Use `popitem()` instead of repeatedly creating an iterator.
* Do not shadow the `id()` builtin
* Drop elements from the dictionary instead of recreating it from scratch (needs some benchmarking)

Reply:
> * using popitem() reverses the order of the iteration, I did not want to do that, though it may be OK.

That's, of course, just a detail to remove the (IMO) hacky refresh of the iterator at every iteration.

> * I had the "purge" how you suggest it, first. But a model has only a handful of fields (not 100k), so I thought there will not be a lot to gain and in the profile, the <dictcomp> only appears almost at the bottom.

If that structure has so few fields, do we even need to keep the while-purge? Can't we just iterate over the items in a simpler for loop? The gains from your changes seem to mostly derive from the change of data structure. How much is due to this purge detail?

Reply:
That said @aj-fuentes resolved this conversation, so he will probably share something of higher relevance on the matter.

Reply:
> The pruning part seems broken:

That's why I didn't post it here, and rather just marked as resolved. OTOH this works
`for other_field in list(fields_recs):` the extra list should add too much overhead.

Reply:
A more representative benchmark:

```python
def f2(s, removal_s):
    for el in removal_s:
        try:
            s.remove(el)
        except KeyError:
            continue

def g2(s, removal_s):
    for el in removal_s:
        if el in s:
            s.remove(el)

def h2(s, removal_s):
    for el in removal_s:
        with contextlib.suppress(KeyError):
            # although we can't quite do this
            s.remove(el)


# most fail
s = set(range(1000))
removal_s = set(range(990, 2000))
print(timeit(lambda: f2(s, removal_s), number=1000))    # 0.09494097800052259
print(timeit(lambda: g2(s, removal_s), number=1000))    # 0.012841757998103276
print(timeit(lambda: h2(s, removal_s), number=1000))    # 0.327394963002007

# most succeed
s = set(range(1000))
removal_s = set(range(10, 1010))
print(timeit(lambda: f2(s, removal_s), number=1000))    # 0.12158733799878974
print(timeit(lambda: g2(s, removal_s), number=1000))    # 0.03167118099372601
print(timeit(lambda: h2(s, removal_s), number=1000))    # 0.3627551410027081

# 50-50
s = set(range(1000))
removal_s = set(range(500, 1500))
print(timeit(lambda: f2(s, removal_s), number=1000))    # 0.10662447399954544
print(timeit(lambda: g2(s, removal_s), number=1000))    # 0.022108470999228302
print(timeit(lambda: h2(s, removal_s), number=1000))    # 0.34070394599984866
```

Reply:
> OTOH this works `for other_field in list(fields_recs):` the extra list should add too much overhead.

I'm still not sold on whether we really need the pruning part. That said, that's also the only solution I could find to avoid the comprehension.

Reply:
Example: on a model with 20 fields in cache, in only _one_ record we write on 20 fields, for the rest we wrote in only one field. Instead of 20 * 20 (without pruning) we iterate first on 20 for first run then on 1 for the the rest -- if we use prune

Reply:
Mainly, I added the "prune" because of the "other_field" loop, because it is nested within the loop over ids (large factor).

Reply:
Hm yes, that's valid. Especially if we do in the `other_field` loop, like in Alvaro's suggestion!

Reply:
Oh, and that is also why I did the prune outside of the loop over ids. But I will test and benchmark all your suggestions, but I need to do it one-by-one to get meaningful comparisons. This will take more time.